src/io/parquet/reader.rs

```diff
@@ -128,6 +134,37 @@ pub fn read_geoparquet<R: ChunkReader + 'static>(
     )
 }
 
+pub async fn read_geoparquet_async<T: AsyncFileReader + Send + 'static>(
+    input: T,
+    options: GeoParquetReaderOptions,
+) -> Result<GeoTable> {
+    let builder = ParquetRecordBatchStreamBuilder::new(input)
+        .await
+        .unwrap()
+        .with_batch_size(options.batch_size);
+
+    let (arrow_schema, geometry_column_index, target_geo_data_type) = {
+        let parquet_meta = builder.metadata();
+        let arrow_schema = builder.schema().clone();
+        let (geometry_column_index, target_geo_data_type) = parse_geoparquet_metadata(
+            parquet_meta.file_metadata(),
+            &arrow_schema,
+            options.coord_type,
+        )?;
+        (arrow_schema, geometry_column_index, target_geo_data_type)
+    };
+
+    let stream: ParquetRecordBatchStream<T> = builder.build().unwrap();
+    let batches: Vec<RecordBatch> = stream.try_collect::<Vec<_>>().await.unwrap();
```

Comment:
I agree it can be a confusing error message that is hard to debug. In this case, I went to the docs of `ParquetRecordBatchStream` and looked at [its implementation of `Stream`](https://docs.rs/parquet/latest/parquet/arrow/async_reader/struct.ParquetRecordBatchStream.html#impl-Stream-for-ParquetRecordBatchStream%3CT%3E). In that docstring, the bound on `T` is `T: AsyncFileReader + Unpin + Send + 'static`, so the inference is that the bound on `T` in this function needs to additionally have `Unpin`, which indeed fixed the compilation issue.

(I don't actually know what Unpin does, since I don't do a lot of async. If you'd like to learn more, I'd suggest this deep dive: https://fasterthanli.me/articles/pin-and-suffering)

Reply:
Ah cool, thanks for the explanation! Was banging my head trying to figure out what was missing. Now I know more about trait bounds (and how to read the docs a bit better to find the traits) :student: